### PR TITLE
Test download location of proof parameter files.

### LINKED ISSUE #1636: Test download location of proof parameter files.
**Issue summary**

The `forest params-fetch` command should download parameter files to `%DATA/filecoin-proof-parameters` OR to the directory given by the `FIL_PROOFS_PARAMETER_CACHE` environment variable. We should have two tests that verifies both behaviors. This does require network access, though. If downloading the parameter files makes the automated testing too slow, make a note in this issue.

**Other information and links**

### PR DESCRIPTION
**Pull request by tyshko5**
**Summary of changes**
Changes introduced in this pull request:
- 

**Reference issue to close (if applicable)**

Closes  #1636

**Other information and links**

### CODE REVIEW

**Review by lemmih — CHANGES_REQUESTED** [review]
Make sure the PR only includes the changes you want it to include. Right now, it has been mixed up with a bunch of other changes.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: forest/tests/config_tests.rs [hunk 1]
```diff
@@ -121,3 +122,77 @@ fn test_config_env_var() {
 
     assert!(expected_config == actual_config);
 }
+
+#[test]
+fn test_download_location_of_proof_parameter_files_env() {
+    let mut rng = rand::thread_rng();
+    let expected_config = Config {
```
[the remark this answers is not in the document]
Comment (tyshko5) [reply forest/tests/config_tests.rs]: Fixed

File: forest/tests/config_tests.rs [hunk 2]
```diff
@@ -121,3 +122,77 @@ fn test_config_env_var() {
...
+        client: Client {
+            metrics_address: SocketAddr::from_str(&format!("127.0.0.1:{}", rng.gen::<u16>()))
+                .unwrap(),
+            rpc_token: Some("Azazello".into()),
+            genesis_file: Some("cthulhu".into()),
```
[the remark this answers is not in the document]
Comment (tyshko5) [reply forest/tests/config_tests.rs]: Removed

File: forest/tests/config_tests.rs [hunk 3]
```diff
@@ -121,3 +122,77 @@ fn test_config_env_var() {
...
+
+    assert!(list_files.count() > 0);
+}
+
+#[test]
+fn test_download_location_of_proof_parameter_files_default() {
+    let tmp_dir = TempDir::new().unwrap();
+    let mut tmp_dir_path_buf = tmp_dir.path().to_path_buf();
+    let mut rng = rand::thread_rng();
+    let expected_config = Config {
```
[the remark this answers is not in the document]
Comment (tyshko5) [reply forest/tests/config_tests.rs]: Fixed

File: forest/tests/config_tests.rs [hunk 4]
```diff
@@ -121,3 +122,77 @@ fn test_config_env_var() {
...
+        client: Client {
+            data_dir: tmp_dir_path_buf.clone(),
+            metrics_address: SocketAddr::from_str(&format!("127.0.0.1:{}", rng.gen::<u16>()))
+                .unwrap(),
+            rpc_token: Some("Azazello".into()),
+            genesis_file: Some("cthulhu".into()),
```
[the remark this answers is not in the document]
Comment (tyshko5) [reply forest/tests/config_tests.rs]: Removed

File: forest/tests/config_tests.rs [hunk 5]
```diff
@@ -121,3 +122,60 @@ fn test_config_env_var() {
...
 }
+
+#[test]
+fn test_download_location_of_proof_parameter_files_env() {
+    let config = Config::default();
+
+    let mut config_file = tempfile::Builder::new().tempfile().unwrap();
+    config_file
+        .write_all(toml::to_string(&config).unwrap().as_bytes())
+        .expect("Failed writing configuration!");
```
Comment (lemmih) [inline forest/tests/config_tests.rs:133]: This test will still work if you don't use a configuration file, right?

File: forest/tests/config_tests.rs [hunk 6]
```diff
@@ -121,3 +122,60 @@ fn test_config_env_var() {
...
+
+    let list_files = read_dir(tmp_dir.path()).unwrap();
+
+    assert!(list_files.count() > 0);
+}
+
+#[test]
+fn test_download_location_of_proof_parameter_files_default() {
+    let tmp_dir = TempDir::new().unwrap();
+    let mut tmp_dir_path_buf = tmp_dir.path().to_path_buf();
```
Comment (lemmih) [inline forest/tests/config_tests.rs:154]:
```suggestion
    let tmp_param_dir = tmp_dir.path().join("filecoin-proof-parameters");
```

File: forest/tests/config_tests.rs [hunk 7]
```diff
@@ -121,3 +122,60 @@ fn test_config_env_var() {
...
+    let config = Config {
+        client: Client {
+            data_dir: tmp_dir_path_buf.clone(),
```
Comment (lemmih) [inline forest/tests/config_tests.rs:157]:
```suggestion
            data_dir: tmp_dir.path().to_path_buf(),
```

File: forest/tests/config_tests.rs [hunk 8]
```diff
@@ -121,3 +122,60 @@ fn test_config_env_var() {
...
+
+    Command::cargo_bin("forest")
+        .unwrap()
+        .env("FOREST_CONFIG_PATH", config_file.path())
+        .arg("fetch-params")
+        .arg("--keys")
+        .assert()
+        .success();
+
+    tmp_dir_path_buf.push("filecoin-proof-parameters");
```
Comment (lemmih) [inline forest/tests/config_tests.rs:176]:
```suggestion
```

File: forest/tests/config_tests.rs [hunk 9]
```diff
@@ -121,3 +122,60 @@ fn test_config_env_var() {
...
+
+    let list_files = read_dir(tmp_dir_path_buf).unwrap();
```
Comment (lemmih) [inline forest/tests/config_tests.rs:178]:
```suggestion
    let list_files = read_dir(tmp_param_dir).unwrap();
```

File: forest/tests/config_tests.rs (hunk 8 above)
Comment (tyshko5) [reply forest/tests/config_tests.rs:176]: I don't understand this

Comment (lemmih) [reply forest/tests/config_tests.rs:176]: Change the code such that `tmp_dir_path_buf` is not mutable.

**Review by lemmih — APPROVED** [review]

File: forest/tests/config_tests.rs [hunk 10]
```diff
@@ -121,3 +122,50 @@ fn test_config_env_var() {
...
+        .unwrap()
+        .env("FIL_PROOFS_PARAMETER_CACHE", tmp_dir.path())
+        .arg("fetch-params")
+        .arg("--keys")
+        .assert()
+        .success();
+
+    let list_files = read_dir(tmp_dir.path()).unwrap();
+
+    assert!(list_files.count() > 0);
```
Comment (LesnyRumcajs) [inline forest/tests/config_tests.rs:140]: probably `!list_files.is_empty()` would be more idiomatic here and in the second test

**Review by LesnyRumcajs — APPROVED** [review]

**Comment by lemmih** [comment]
@tyshko5 It's up to you whether you want to apply @LesnyRumcajs's suggestion. When you feel it's ready, you can merge this PR on your own.

File: forest/tests/config_tests.rs (hunk 10 above)
Comment (tyshko5) [reply forest/tests/config_tests.rs:140]: There is no such method for that type

**Pull request merged**